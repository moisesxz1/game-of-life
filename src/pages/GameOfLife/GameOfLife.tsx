import React, { useState, useEffect, useCallback } from "react";
import { BOARD_SIZE } from "../../constants";
import { Board } from "../../types";
import { generateEmptyBoard, getNextBoardState } from "../../utils/utils";
import Cell from "../../components/Cell/Cell";
import { throttle } from "lodash";

const GameOfLife: React.FC = () => {
  const [board, setBoard] = useState<Board>(generateEmptyBoard());
  const [isPlaying, setIsPlaying] = useState(false);
  const [speed] = useState(1000);
  const [steps, setSteps] = useState(1);

  const toggleCellState = useCallback((x: number, y: number) => {
    setBoard((prevBoard) =>
      prevBoard.map((row, rowIndex) =>
        row.map((cell, colIndex) => {
          if (rowIndex === x && colIndex === y) {
            return !cell;
          }
          return cell;
        })
      )
    );
  }, []);

  const nextState = throttle(() => {
    setBoard((prevBoard) => getNextBoardState(prevBoard));
  }, 100); // Only allow it to be called every 100ms

  const togglePlay = () => {
    setIsPlaying(!isPlaying);
  };

  const advanceXSteps = () => {
    let newBoard = board;
    for (let i = 0; i < steps; i++) {
      newBoard = getNextBoardState(newBoard);
    }
    setBoard(newBoard);
  };

  useEffect(() => {
    if (isPlaying) {
      const interval = setInterval(() => {
        nextState();
      }, speed);
      return () => clearInterval(interval);
    }
  }, [isPlaying, speed]);

  return (
    <div>
      <div
        style={{
          display: "grid",
          gridTemplateColumns: `repeat(${BOARD_SIZE}, 20px)`,
        }}
      >
        {board.map((row, rowIndex) =>
          row.map((cell, colIndex) => (
            <Cell
              key={`${rowIndex}-${colIndex}`}
              onClick={() => toggleCellState(rowIndex, colIndex)}
              isAlive={cell}
            />
          ))
        )}
      </div>

      <button onClick={nextState}>Next State</button>
      <button onClick={togglePlay}>{isPlaying ? "Pause" : "Play"}</button>
      <input
        type="number"
        value={steps}
        onChange={(e) => setSteps(Number(e.target.value))}
      />
      <button onClick={advanceXSteps}>Advance {steps} Steps</button>
    </div>
  );
};

export default GameOfLife;
