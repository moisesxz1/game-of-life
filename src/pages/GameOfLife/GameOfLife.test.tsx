import { render, screen, fireEvent } from "@testing-library/react";
import GameOfLife from "./GameOfLife";

describe("GameOfLife Component", () => {
  it("matches the snapshot", () => {
    const { asFragment } = render(<GameOfLife />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the grid with cells", () => {
    render(<GameOfLife />);
    const gridSize = 400;

    // Check if grid cells are rendered
    const cells = screen.getAllByTestId("cell");
    expect(cells.length).toBe(gridSize);
  });

  it("toggles cell state when clicked", () => {
    render(<GameOfLife />);

    const cell = screen.getAllByTestId("cell")[0];

    expect(cell).toHaveClass("cell is-dead");

    fireEvent.click(cell);

    // Cell should now be alive (black background)
    expect(cell).toHaveClass("cell is-alive");
  });

  it("Should start and stop the game when play/pause is clicked", () => {
    jest.useFakeTimers();

    render(<GameOfLife />);

    const playButton = screen.getByText("Play");
    fireEvent.click(playButton);

    expect(playButton.textContent).toBe("Pause");

    jest.advanceTimersByTime(1000);

    // Assuming the game updates the board state every second, check if the board has changed
    const cells = screen.getAllByTestId("cell");
    const firstCell = cells[0];

    jest.advanceTimersByTime(1000);
    expect(firstCell).toHaveClass("cell is-dead");

    fireEvent.click(playButton);
    expect(playButton.textContent).toBe("Play");

    jest.clearAllTimers();
  });
});
