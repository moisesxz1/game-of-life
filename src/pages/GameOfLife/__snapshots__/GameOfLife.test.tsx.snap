// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameOfLife Component matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      style="display: grid; grid-template-columns: repeat(20, 20px);"
    >
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
      <div
        class="cell is-dead"
        data-testid="cell"
      />
    </div>
    <button>
      Next State
    </button>
    <button>
      Play
    </button>
    <input
      type="number"
      value="1"
    />
    <button>
      Advance 1 Steps
    </button>
  </div>
</DocumentFragment>
`;
